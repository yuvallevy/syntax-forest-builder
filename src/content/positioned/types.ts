import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type Width = number;

export type PlotCoords = {
  plotX: PlotX;
  plotY: PlotY;
};

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: Id[];
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  slice: StringSlice;
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedStrandedNode = NodeCommon & WithPositionInTree;

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode | PositionedStrandedNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: PlotCoords;
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
