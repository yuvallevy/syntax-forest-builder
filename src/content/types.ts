export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;

export type StringSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type NodeIndicatorInPlot = {
  treeId: Id;
  nodeId: Id;
};

export type TreeCommon = {
  sentence: Sentence;
}

export type NodeCommon = {
  label: NodeLabel;
};
