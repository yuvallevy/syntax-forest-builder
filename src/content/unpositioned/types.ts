import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;

export type PlotCoordsOffset = {
  dPlotX: DPlotX;
  dPlotY: DPlotY;
};

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: PlotCoordsOffset;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: Id[];
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: StringSlice;
  triangle: boolean;
};

export type UnpositionedStrandedNode = NodeCommon & WithOffsetInTree & {
  formerDescendants?: IdMap<UnpositionedNode>;
  formerSlice?: StringSlice;
  formerlyTriangle?: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode | UnpositionedStrandedNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export const isTerminal = (node: UnpositionedNode): node is UnpositionedTerminalNode =>
  'slice' in node;
