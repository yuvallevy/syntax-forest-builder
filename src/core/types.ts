export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;
type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;
type Width = number;

export type NodeSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type UnpositionedForest = { plots: IdMap<UnpositionedPlot>; };

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

type TreeCommon = {
  sentence: Sentence;
}

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: { dPlotX: DPlotX; dPlotY: DPlotY; };
};

export type NodeCommon = {
  label: NodeLabel;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: IdMap<UnpositionedNode>;
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: NodeSlice;
  triangle: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: IdMap<PositionedNode>;
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: {
    plotX: PlotX;
    plotY: PlotY;
  };
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
