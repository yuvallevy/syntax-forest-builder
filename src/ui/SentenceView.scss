.SentenceView-input {
  position: absolute;
  padding: 0;
  border: 0;
  font-size: 16px;
  font-family: 'Times New Roman', Times, serif;

  &:focus {
    outline: unset;
  }
}
