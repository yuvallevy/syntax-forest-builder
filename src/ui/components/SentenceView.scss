@import "PlotView";

.SentenceView--input {
  position: absolute;
  padding: 0;
  border: 0;
  font-size: 16px;
  font-family: $plotViewFontFamily;

  &:focus {
    outline: unset;
  }

  &.box-selecting {
    z-index: -100;
  }
}
