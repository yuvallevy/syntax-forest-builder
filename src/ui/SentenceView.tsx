import { useRef } from 'react';
import { StringSlice, PositionedTree, Sentence, Id } from '../core/types';
import './SentenceView.scss';

// A tree with no sentence will take up this width instead of 0 (or something close to 0):
const EMPTY_SENTENCE_WIDTH = 120;

// This will be added to the width of an input field determined by its content width, in case the width calculation is
// inaccurate for whatever reason
// (it doesn't allow placing one tree exactly at the end of another one, but that's not a common occurrence)
const EXTRA_SENTENCE_WIDTH = 4;

interface SentenceViewProps {
  tree: PositionedTree;
  treeId: Id;
  onChange: (newSentence: Sentence, oldSelectedSlice: StringSlice) => void;
  onSelect: (slice: StringSlice) => void;
  onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
}

const getSelectionSlice = (element: HTMLInputElement): StringSlice | null =>
  element.selectionStart !== null && element.selectionEnd !== null
    ? [element.selectionStart, element.selectionEnd]
    : null;

const SentenceView: React.FC<SentenceViewProps> = ({ tree, treeId, onChange, onSelect, onKeyDown }) => {
  // Keep track of the previous selection so we can report it whenever a change is made
  // (the input event only carries information about the *new* selection, hence this hack)
  const oldSelection = useRef<StringSlice | null>(null);

  return <input
    type="text"
    id={treeId}
    value={tree.sentence}
    className="SentenceView-input"
    style={{
      left: tree.position.plotX,
      top: tree.position.plotY,
      width: tree.sentence.length === 0 ? EMPTY_SENTENCE_WIDTH : tree.width + EXTRA_SENTENCE_WIDTH,
    }}
    placeholder="Type a sentence..."
    onInput={e => onChange(e.currentTarget.value,
      oldSelection.current || [e.currentTarget.value.length, e.currentTarget.value.length])}
    onSelect={e => {
      const slice = getSelectionSlice(e.currentTarget);
      if (slice) onSelect(slice);
      oldSelection.current = getSelectionSlice(e.currentTarget);
    }}
    onKeyDown={onKeyDown}
  />;
};

export default SentenceView;
