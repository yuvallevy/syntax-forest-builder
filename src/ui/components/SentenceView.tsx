import { useRef } from 'react';
import { StringSlice, Sentence, Id } from '../../content/types';
import { PositionedTree } from '../../content/positioned/types';
import './SentenceView.scss';

// A tree with no sentence will take up this width instead of 0 (or something close to 0):
const EMPTY_SENTENCE_WIDTH = 120;

// This will be added to the width of an input field determined by its content width, in case the width calculation is
// inaccurate for whatever reason
// (it doesn't allow placing one tree exactly at the end of another one, but that's not a common occurrence)
const EXTRA_SENTENCE_WIDTH = 4;

interface SentenceViewProps {
  tree: PositionedTree;
  treeId: Id;
  className?: string;
  onBlur: (event: React.FocusEvent<HTMLInputElement>) => void;
  onChange: (newSentence: Sentence, oldSelectedSlice: StringSlice) => void;
  onSelect: (slice: StringSlice) => void;
  onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
}

const getSelectionSlice = (element: HTMLInputElement): StringSlice | null =>
  element.selectionStart !== null && element.selectionEnd !== null
    ? [element.selectionStart, element.selectionEnd]
    : null;

const SentenceView: React.FC<SentenceViewProps> = ({
  tree,
  treeId,
  className,
  onBlur,
  onChange,
  onSelect,
  onKeyDown,
}) => {
  // Keep track of the previous selection so we can report it whenever a change is made
  // (the input event only carries information about the *new* selection, hence this hack)
  const oldSelection = useRef<StringSlice | null>(null);

  return <input
    type="text"
    id={treeId}
    value={tree.sentence}
    className={'SentenceView--input' + (className ? ` ${className}` : '')}
    style={{
      left: tree.position.plotX,
      top: tree.position.plotY,
      width: tree.sentence.length === 0 ? EMPTY_SENTENCE_WIDTH : tree.width + EXTRA_SENTENCE_WIDTH,
    }}
    placeholder="Type a sentence..."
    onBlur={onBlur}
    onInput={e => onChange(e.currentTarget.value,
      oldSelection.current || [e.currentTarget.value.length, e.currentTarget.value.length])}
    onSelect={e => {
      const slice = getSelectionSlice(e.currentTarget);
      if (slice) onSelect(slice);
      oldSelection.current = getSelectionSlice(e.currentTarget);
    }}
    onKeyDown={onKeyDown}
  />;
};

export default SentenceView;
