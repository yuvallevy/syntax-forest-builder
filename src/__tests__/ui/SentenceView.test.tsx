import { render } from '@testing-library/react';
import { PositionedTree } from '../../core/types';
import SentenceView from '../../ui/SentenceView';

describe('sentence rendering', () => {
  const tree: PositionedTree = {
    sentence: 'Noun verb phrases.',
    nodes: {},
    position: { plotX: -100, plotY: 20 },
    width: 123,
  };

  it('renders an input box with the sentence', () => {
    expect(render(
      <SentenceView tree={tree} onChange={() => null} onSelect={() => null} />
    ).asFragment()).toMatchSnapshot();
  });
});
