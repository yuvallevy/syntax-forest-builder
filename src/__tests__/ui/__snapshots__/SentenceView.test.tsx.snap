// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sentence rendering renders an input box with the sentence 1`] = `
<DocumentFragment>
  <input
    class="SentenceView-input"
    id="t123456"
    placeholder="Type a sentence..."
    style="left: -100px; top: 20px; width: 127px;"
    type="text"
    value="Noun verb phrases."
  />
</DocumentFragment>
`;
